### ISSUE TITLE
Tail input plugin doesn't work for named pipes (FIFOs)

### ISSUE THREAD

**Issue by ccampo133** [fluent/fluent-bit#6829]
## Bug Report

**Describe the bug**
When using a named pipe (FIFO) as the file for the `tail` input plugin, Fluent Bit starts successfully without errors, but never processes any input data.

**To Reproduce**

Create the pipe:

```
$ mkfifo /path/to/log.fifo
```

Example `fluent-bit.conf`:

```
[INPUT]
    Name tail
    Path /path/to/log.fifo

[OUTPUT]
    Name stdout
```

Then start fluent-bit:

```
$ fluent-bit -c fluent-bit.conf -f 1
Fluent Bit v2.0.8
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/02/10 15:42:40] [ info] [fluent bit] version=2.0.8, commit=, pid=25834
[2023/02/10 15:42:40] [ info] [storage] ver=1.4.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/02/10 15:42:40] [ info] [cmetrics] version=0.5.8
[2023/02/10 15:42:40] [ info] [ctraces ] version=0.2.7
[2023/02/10 15:42:40] [ info] [input:tail:tail.0] initializing
[2023/02/10 15:42:40] [ info] [input:tail:tail.0] storage_strategy='memory' (memory only)
[2023/02/10 15:42:40] [ info] [sp] stream processor started
[2023/02/10 15:42:40] [ info] [output:stdout:stdout.0] worker #0 started
```

In another terminal, write to the pipe:

```
$ echo "Hello world" > /path/to/log.fifo
```

**Expected behavior**
At this point I would expect Fluent Bit to process the message `Hello world`, however it doesn't do anything and `echo` is still blocked writing to the pipe. I expected this to work because "everything is a file", but presumably Fluent Bit expects regular files vs FIFOs.

**Your Environment**

* Version used: v2.0.8
* Operating System and version: MacOS Ventura 13.2, but tested on Linux as well with the same issue.
* Filters and plugins: `tail` input

**Additional context**

The use-case here is for applications that want to redirect their output to Fluent Bit _without writing to disk_.

Looks like this may have been addressed by https://github.com/fluent/fluent-bit/pull/1725, but it was closed without being fully vetted.

**Comment by ccampo133** [comment]
Created a [PR](https://github.com/fluent/fluent-bit/pull/6831) for this, based on the earlier PR https://github.com/fluent/fluent-bit/pull/1725.

It _appears_ to work for me locally using the configuration described above, but see my PR for the disclaimers :).

**Comment by sharvil** [comment]
I'm very interested in seeing the PR merged and this issue resolved. Without named pipe support, we're stuck writing unnecessarily to disk and rotating/deleting content periodically which adds complexity to our deployment.

**Comment by leonardo-albertovich** [comment]
I'm not really sure if that PR will be merged but for the moment you can use the `exec` input plugin in threaded mode running your favorite tool to read from the FIFO.

If you do this, it's key that you run the input plugin in threaded mode because `exec` runs in a busy loop ingesting lines until the child process exits so it would block the main thread otherwise.

Hope it helps.

**Comment by ccampo133** [comment]
A FIFO is a pretty well understood type of file. Considering `tail -f` works as expected on FIFOs, I think it is reasonable to expect this plugin to work with with them as well. My PR was just a few simple changes to make it work, and didn't add considerable complications to the plugin. I hope that it can be reviewed and merged.

The problem with using a `exec` plugin is that they are only ever triggered on a timer, rather than in response to events (see https://github.com/fluent/fluent-bit/blob/master/plugins/in_exec/in_exec.c#L245-L260). This is insufficient for a lot of cases. I opened an issue in fluent-bit-go (https://github.com/fluent/fluent-bit-go/issues/59) which describes a similar deficiency with the Go plugins. If the `exec` plugin (and Go plugins while we're at it) could respond to events in addition to a timer, that would pretty helpful as well... but that sounds like a different issue altogether.

**Comment by leonardo-albertovich** [comment]
Don't worry, that's just my point of view, not my call. As for the `exec` plugin, could you clarify what does it lack feature wise?

Just yesterday I showed someone else how to use it to keep a long running command producing input in a background thread so wouldn't this example achieve what you are trying to? 

```
[SERVICE]
    flush 1
    grace 1
    log_level info

[INPUT]
    name dummy
    dummy {"log": "this runs in the main thread"}
    rate 1

[INPUT]
    # This runs in a different thread and does not block the rest of the system
    name exec
    command tail -n 0 -f /path/to/fifo
    threaded on
    oneshot on

[OUTPUT]
    name stdout
    match *
```

You  could remove the `oneshot` option in case you would want tail to run again if the process ends for any reasons.

To be clear, I'm not a fan of these "run it as a command" approach, I think it's not appropriate but I'm not entirely convinced by the suggested modification mostly due to how convoluted the `tail` input plugin already is and would like to hold back a little bit on it if possible.

**Comment by leonardo-albertovich** [comment]
Well, the key part in this approach is to run the input plugin in threaded mode which was released in fluent-bit 2.0 but I think we need to improve the documentation for it since I found it to be a bit lacking when I tried to look for a reference to share.

As for the PR, I think we will be able to pay some attention to it soon, can't promise anything but I agree with you about it being something we should support natively, I just want to ensure that we don't make a mistake.
